core: retry part 5, add RetryPolicy data object

dapengzhang0 opened the pull request:
The logic of using retry policy to compute retry plan will be in one of upcoming PRs.

dapengzhang0 commented:
Rebased.

thelinuxfoundation commented:
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* One or more commits are not associated with a GitHub user. The email used in the Git commit must match a verified email address in a GitHub user profile. For more information, please see https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/

_Regards,
CLA GitHub bot_

ericgribkoff commented on core/src/main/java/io/grpc/internal/RetriableStream.java:
```diff
@@ -777,4 +785,58 @@ public long addAndGet(long newBytesUsed) {
+      this.initialBackoffInSeconds = initialBackoffInSeconds;
+      checkArgument(
+          maxBackoffInSeconds >= initialBackoffInSeconds,
+          "maxBackoofInSeconds should be at least initialBackoffInSeconds");
```
s/Backoof/Backoff

dapengzhang0 replied:
Done

ericgribkoff commented on core/src/main/java/io/grpc/internal/RetriableStream.java:
```diff
@@ -777,4 +785,58 @@ public long addAndGet(long newBytesUsed) {
+    public int hashCode() {
+      return Objects.hashCode(
+          maxAttempts, initialBackoffInSeconds, maxBackoffInSeconds, backoffMultiplier,
+                              retryableStatusCodes);
```
Indentation seems off here?

dapengzhang0 replied:
Done.

ericgribkoff commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -1068,6 +1096,104 @@ public void run() {
+      public RetryPolicy get(MethodDescriptor<?, ?> method) {
+        RetryPolicy retryPolicy = fullMethodNameMap.get(method.getFullMethodName());
+        if (retryPolicy == null) {
+          retryPolicy = serviceNameMap
```
Add test coverage for this fallback logic?

dapengzhang0 replied:
I think this test code covers it

```java
assertEquals(
          new RetryPolicy(
              3, parseDouble("2.1"), parseDouble("2.2"), parseDouble("3"),
              Arrays.asList(Code.UNAVAILABLE, Code.RESOURCE_EXHAUSTED)),
          retryPolicies.get(builder.setFullMethodName("SimpleService1/not_exist").build()));
```

ericgribkoff commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -1068,6 +1096,104 @@ public void run() {
+    return getRetryPolicies(serviceConfig);
+  }
+
+  // TODO(zdapeng): maybe move it to a more appropriate place
```
+1 (but optional, because I'm not sure where would be more appropriate, but it seems like it might be nice to keep the json parsing in a separate location)

ericgribkoff commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -1068,6 +1096,104 @@ public void run() {
+            int maxAttempts = retryPolicy.getAsJsonPrimitive("maxAttempts").getAsInt();
+            String initialBackoffStr =
+                retryPolicy.getAsJsonPrimitive("initialBackoff").getAsString();
+            checkState(
```
These should be checkArgument(), shouldn't they?

dapengzhang0 replied:
My understanding is `checkArgument` (`/** Ensures the truth of an expression involving one or more parameters to the calling method */`) should check the argument (here is `serviceConfig`) directly in the expression, but here we check `initialBackoffStr` which is an intermediate object so `checkState` is more suitable. I'm not sure my understanding is correct through.

ericgribkoff commented on core/src/main/java/io/grpc/internal/RetriableStream.java:
```diff
@@ -777,4 +785,58 @@ public long addAndGet(long newBytesUsed) {
+    private final int maxAttempts;
+    private final double initialBackoffInSeconds;
+    private final double maxBackoffInSeconds;
+    private final double backoffMultiplier;
```
backoffMultiplier should be an int

dapengzhang0 replied:
That surprises me. Why I can not set backoffMultiplier = 1.5 if I consider 2 is too large?

ericgribkoff replied:
The schema from the design (and quoted in this PR in the comment block in `ManagedChannelImpl#getRetryPolicies`) says: `"backoffMultiplier": number`. I'm not aware of any strong constraint that the design couldn't be changed to allow floating point numbers here, but neither did a strong (or any) need for this come up during the design phase, so it's currently specified as an integer.

ericgribkoff replied:
Nevermind, I was incorrect - I hadn't realized json numbers can be either float or int. We're using the same json type for maxAttempts (which must be an integer) as for backoffMultiplier, but in the proto representation backoffMultiplier can indeed be a floating point number.

dapengzhang0 commented:
Moved out the gson logic in ManagedChannelImpl and leave it unimplemented for the moment. Moved the gson logic into test because in test it will still be helpful when I have a valid RetryPolicy.

ejona86 commented on core/src/test/java/io/grpc/internal/RetryPolicyTest.java:
```diff
@@ -0,0 +1,188 @@
+            JsonArray retryableStatusCodes = retryPolicy.getAsJsonArray("retryableStatusCodes");
+            Set<Code> codeSet = new HashSet<Code>(retryableStatusCodes.size());
+            for (JsonElement retryableStatusCode : retryableStatusCodes) {
+              codeSet.add(Code.valueOf(retryableStatusCode.getAsString()));
```
Should unknown statuses be ignored? I think so. That should probably be explicitly stated in the spec.

dapengzhang0 replied:
+ @ericgribkoff if the spec need be updated

dapengzhang0 commented on core/src/main/java/io/grpc/internal/RetriableStream.java:
```diff
@@ -80,14 +86,16 @@
     this.callExecutor = callExecutor;
     this.scheduledExecutorService = scheduledExecutorService;
     this.headers = headers;
+    this.retryPolicy = checkNotNull(retryPolicy);
```
Done.

dapengzhang0 commented on core/src/main/java/io/grpc/internal/RetriableStream.java:
```diff
@@ -777,4 +785,57 @@ public long addAndGet(long newBytesUsed) {
+
+    /** No retry. */
+    static final RetryPolicy DEFAULT =
+        new RetryPolicy(1, 0, 0, 1, Collections.<Code>emptyList());
```
Done.

dapengzhang0 commented on core/src/main/java/io/grpc/internal/RetriableStream.java:
```diff
@@ -777,4 +785,57 @@ public long addAndGet(long newBytesUsed) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
```
Done.
